staticjinja/dep_graph.py

```diff
@@ -0,0 +1,127 @@
+# -*- coding:utf-8 -*-
+
+"""
+Dependency graph for staticjinja
+"""
+
+
+class DepGraph(object):
+    """
+    A directed graph which will handle dependencies between sources in a site.
+
+    :param parents:
+        A dictionary whose keys are vertices of the graph and each value is the
+        set of child vertices of the key vertex.
+
+    :param children:
+        A dictionary whose keys are vertices of the graph and each value is the
+        set of parents vertices of the key vertex.
+
+    """
+    def __init__(self, parents, children):
+        self.parents = parents
+        self.children = children
+
+    @classmethod
+    def from_parents(cls, parents):
+        """
+        Buils a graph from a dictionnary of directed edges
+        by building the dictionnary of children edges.
+
+        :param parents:
+            A dictionary whose keys are vertices of the graph and each value is
+            the set of child vertices of the key vertex.
+        """
+        children = {}
+        for vertex in parents:
+            children[vertex] = set()
+        for vertex in parents:
+            for d in parents[vertex]:
+                children[d].add(vertex)
+        return cls(parents, children)
+
+    def connected_components(self, direction, start):
+        """Returns the (directed) connected component of start in the graph
+        (including start).
+
+        Uses a classical depth first search.
+
+        :param direction: either 'descendants' or 'ancestors'
+
+        :param start: the vertex in the graph whose connected component we
+            seek.
+
+        """
+
+        if direction == 'descendants':
+            adjacency = self.children
+        elif direction == 'ancestors':
+            adjacency = self.parents
+        else:
+            raise ValueError(
+                    'direction should be either descendants or ancestors'
+                    )
```

Comment:
Sorry, I must not have been clear-- the difference with my proposal and the current version is that if we need to change the signatures of either function in the code I proposed, it's trivial, but if we need to change it in the current version then we're going to run into a problem because we'll have something like this:

```
def gen_connected_components(self, direction, start, arg1, arg2):
    # parse arg1 and arg2 based on the value of direction
    # ...
```

``range()`` for instance is a logical strength function. The meaning of its args change based on how many there are.